### PR TITLE
Feature/demo adjustments

### PR DESCRIPTION
**Pull request by rttwente**
Commented code. Added nested switch statements and removed all documentation

### CODE REVIEW

**Comment by HaicoDorenbos** [comment]
> Commented code. Added nested switch statements and removed all documentation

:scream:

File: src/stp/new_plays/GenericPass.cpp [hunk 1]
```diff
@@ -19,14 +19,14 @@
...
-    keepPlayInvariants.emplace_back(std::make_unique<invariant::NormalPlayGameStateInvariant>());
-    keepPlayInvariants.emplace_back(std::make_unique<invariant::FreedomOfRobotsInvariant>());
+//    startPlayInvariants.clear();
+//    startPlayInvariants.emplace_back(std::make_unique<invariant::NormalPlayGameStateInvariant>());
+//    startPlayInvariants.emplace_back(std::make_unique<invariant::BallCloseToUsInvariant>());
+//    startPlayInvariants.emplace_back(std::make_unique<invariant::BallOnOurSideInvariant>());
+//
+//    keepPlayInvariants.clear();
+//    keepPlayInvariants.emplace_back(std::make_unique<invariant::NormalPlayGameStateInvariant>());
+//    keepPlayInvariants.emplace_back(std::make_unique<invariant::FreedomOfRobotsInvariant>());
```
Comment (HaicoDorenbos) [inline src/stp/new_plays/GenericPass.cpp]: Deadcode should be removed, not commented out... If you really think you need it, place it at a different branch. If you think that branch is forgotten then put a reference to that branch on the wiki page.

File: src/stp/new_plays/GenericPass.cpp [hunk 2]
```diff
@@ -41,21 +41,21 @@ GenericPass::GenericPass() : Play() {
...
-uint8_t GenericPass::score(world_new::World* world) noexcept { return 50; }
+uint8_t GenericPass::score(world_new::World* world) noexcept { return 100; }
 
 void GenericPass::calculateInfoForRoles() noexcept {
     // Keeper
     stpInfos["keeper"].setPositionToShootAt(Vector2{0.0, 0.0});
     stpInfos["keeper"].setEnemyRobot(world->getWorld()->getRobotClosestToBall(world_new::them));
 
-    const Vector2 passingPosition = calculatePassLocation();
+    const Vector2 passingPosition = Vector2(-field.getFieldLength()/2, -field.getFieldWidth()/2);//calculatePassLocation();
```
Comment (HaicoDorenbos) [inline src/stp/new_plays/GenericPass.cpp]: No, this type of unclarity & code cloning does not happen under my watch. :disappointed: What you actually should use is: field.getBottomLeftCorner(). In general you should never encode position in this way, but create a separate definition in the field class for these special locations.

File: src/stp/new_plays/GenericPass.cpp [hunk 3]
```diff
@@ -67,17 +67,17 @@ Dealer::FlagMap GenericPass::decideRoleFlags() const noexcept {
...
-    flagMap.insert({"halt_7", {not_important}});
+//    flagMap.insert({"receiver", {receiverFlag}});
+//    flagMap.insert({"halt_0", {not_important}});
+//    flagMap.insert({"halt_1", {not_important}});
+//    flagMap.insert({"halt_2", {not_important}});
+//    flagMap.insert({"halt_3", {not_important}});
+//    flagMap.insert({"halt_4", {not_important}});
+//    flagMap.insert({"halt_5", {not_important}});
+//    flagMap.insert({"halt_6", {not_important}});
+//    flagMap.insert({"halt_7", {not_important}});
```
Comment (HaicoDorenbos) [inline src/stp/new_plays/GenericPass.cpp]: Haico is a little bit disappointed if he sees this kind of stuff :disappointed:, never out comment dead code but remove it instead...

File: src/stp/new_tactics/TestTactic.cpp [hunk 4]
```diff
@@ -28,11 +28,29 @@ void TestTactic::onTerminate() noexcept {
...
+        skillStpInfo.setPositionToMoveTo(Vector2(length / 8, -width / 8));
+        break;
+    case 2:
+        skillStpInfo.setPositionToMoveTo(Vector2(-length / 8, -width / 8));
+        break;
+    case 3:
+        skillStpInfo.setPositionToMoveTo(Vector2(-length / 8, width / 8));
+        break;
+    case 4:
+        skillStpInfo.setPositionToMoveTo(Vector2(length / 8, width / 8));
```
Comment (HaicoDorenbos) [inline src/stp/new_tactics/TestTactic.cpp]: Like I said, position should not get encoded this way. Create a separate variable in the Field class for these positions or create a function in the FieldComputation class that returns these positions instead.

File: src/utilities/Constants.cpp [hunk 5]
```diff
@@ -249,9 +249,9 @@ QColor Constants::SELECTED_ROBOT_COLOR() { return Qt::magenta; }
 
 std::vector<QColor> Constants::TACTIC_COLORS() { return {{255, 0, 255, 50}, {0, 255, 255, 50}, {255, 255, 0, 50}, {0, 255, 0, 50}, {0, 0, 255, 100}}; }
 
-pidVals Constants::standardNumTreePID() { return GRSIM() ? pidVals(3.2, 0.0, 1.4) : pidVals(3.1, 0.0, 0.6); }
+pidVals Constants::standardNumTreePID() { return GRSIM() ? pidVals(2.5, 0.0, 0) : pidVals(3.1, 0.0, 0.6); }
 
-pidVals Constants::standardBasicPID() { return GRSIM() ? pidVals(3.4, 0.0, 0.4) : pidVals(4.0, 0.0, 0.4); }
+pidVals Constants::standardBasicPID() { return GRSIM() ? pidVals(2.5, 0.0, 0) : pidVals(4.0, 0.0, 0.4); }
```
Comment (HaicoDorenbos) [inline src/utilities/Constants.cpp]: Why 2.5? Why not 6.2? Or 1.8? Make sure to document your choices for constant values...

File: src/utilities/IOManager.cpp [hunk 6]
```diff
@@ -97,19 +97,24 @@ const proto::SSL_Referee &IOManager::getRefereeData() {
...
-        if (robot) {
-            if (cmd.kicker()) {
-                interface::Input::drawData(interface::Visual::SHOTLINES, {robot->get()->getPos()}, Qt::green, robot->get()->getId(), interface::Drawing::CIRCLES, 36, 36, 8);
+        if (world_new::World::instance()->getWorld()) {
+            // the geneva cannot be received from world, so we set it when it gets sent.
+            auto robot = world_new::World::instance()->getWorld()->getRobotForId(cmd.id(), true);
+            if (robot) {
+                if (cmd.kicker()) {
+                    interface::Input::drawData(interface::Visual::SHOTLINES, {robot->get()->getPos()}, Qt::green,
+                            robot->get()->getId(), interface::Drawing::CIRCLES, 36, 36, 8);
```
Comment (HaicoDorenbos) [inline src/utilities/IOManager.cpp]: Make sure you use constants for 36 and 8. These are totally undocumented magic values, which will cause huge problems in the future...

**Review by HaicoDorenbos — DISMISSED** [review]
I know what you think, just merge this PR and ignore Haico's feedback. But you introduce a lot of major design issues that are really hazardous for code development...

File: src/stp/new_roles/Passer.cpp [hunk 7]
```diff
@@ -9,6 +9,7 @@
 #include <roboteam_utils/Print.h>
 
 #include <utility>
+#include <include/roboteam_ai/stp/new_tactics/GetBallInDirection.h>
```
Comment (timovdk) [inline src/stp/new_roles/Passer.cpp]:
```suggestion
#include "stp/new_tactics/GetBallInDirection.h"
```

File: src/stp/new_skills/Chip.cpp [hunk 8]
```diff
@@ -15,6 +15,7 @@ Status Chip::onUpdate(const StpInfo &info) noexcept {
     // Set chip command
     command.set_chipper(true);
     command.set_chip_kick_vel(chipVelocity);
+    command.set_chip_kick_forced(true);
```
Comment (timovdk) [inline src/stp/new_skills/Chip.cpp]:
```suggestion
```

File: src/stp/new_tactics/GetBall.cpp [hunk 9]
```diff
@@ -35,11 +35,11 @@ std::optional<StpInfo> GetBall::calculateInfoForSkill(StpInfo const &info) noexc
...
     // the robot will go to the position of the ball
     double ballDistance = (ballPosition - robotPosition).length();
-    Vector2 newRobotPosition = robotPosition + (ballPosition - robotPosition).stretchToLength(ballDistance - stp::control_constants::CENTER_TO_FRONT - stp::control_constants::BALL_RADIUS);
+    Vector2 newRobotPosition = robotPosition + (ballPosition - robotPosition).stretchToLength(ballDistance) - 0.3*control_constants::CENTER_TO_FRONT;
 
     if (ballDistance < control_constants::TURN_ON_DRIBBLER_DISTANCE) {
-        skillStpInfo.setAngle((ballPosition - robotPosition).angle());
-        skillStpInfo.setDribblerSpeed(100);
+        skillInfo.setAngle((ballPosition - robotPosition).angle());
+        skillInfo.setDribblerSpeed(25);
```
Comment (timovdk) [inline src/stp/new_tactics/GetBall.cpp]:
```suggestion
        skillStpInfo.setAngle((ballPosition - robotPosition).angle());
        skillStpInfo.setDribblerSpeed(25);
```

**Review by timovdk — APPROVED** [review]

File: src/stp/new_plays/TimeOut.cpp [hunk 10]
```diff
@@ -29,7 +29,7 @@ uint8_t TimeOut::score(world_new::World* world) noexcept { return 100; }
 
 void TimeOut::calculateInfoForRoles() noexcept {
     const auto xPosition = -4 * control_constants::ROBOT_RADIUS;
-    const auto yPosition = Constants::STD_TIMEOUT_TO_TOP() ? field.getFieldWidth() / 2 : -field.getFieldWidth() / 2;
+    const auto yPosition = Constants::STD_TIMEOUT_TO_TOP() ? field.getFieldWidth() / 2 - 0.15 : -field.getFieldWidth() / 2 + 0.15;
```
Comment (HaicoDorenbos) [inline src/stp/new_plays/TimeOut.cpp]: You are basically hard coding the coordinate system if you use this and it is a type of code cloning. Use field.getTopMostY() and field.getBottomMostY() instead.

File: src/stp/Skill.cpp [hunk 11]
```diff
@@ -84,7 +84,7 @@ void Skill::limitAngularVel() noexcept {
     // Limit the angular velocity when the robot has the ball by setting the target angle in small steps
     // TODO: Might want to limit on the robot itself
     if (robot->hasBall() && command.use_angle()) {
-        double angleRate = 0.2 * M_PI; // Angle increment per tick TODO: TUNE
+        double angleRate = 0.1 * M_PI; // Angle increment per tick TODO: TUNE
```
Comment (HaicoDorenbos) [inline src/stp/Skill.cpp]: Why not make a class constant of this variable?

File: src/stp/new_constants/ControlConstants.cpp [hunk 12]
```diff
@@ -17,7 +17,7 @@ constexpr double MAX_CHIP_POWER = 8;
 constexpr double MIN_CHIP_POWER = 1.01;
 
 // Dribbler constants
-const double TURN_ON_DRIBBLER_DISTANCE = 3 * ROBOT_RADIUS;
+const double TURN_ON_DRIBBLER_DISTANCE = 5 * ROBOT_RADIUS;
```
Comment (HaicoDorenbos) [inline src/stp/new_constants/ControlConstants.cpp]: The naming does not yet explain me the meaning of the variable. So additional documentation is needed here.

File: src/stp/new_constants/ControlConstants.cpp [hunk 13]
```diff
@@ -36,11 +36,11 @@ constexpr double MAX_DRIBBLER_CMD = 31;
 
 // HasBall margins
 constexpr double HAS_BALL_ANGLE_ERROR_MARGIN = 0.106;
-constexpr double HAS_BALL_DISTANCE_ERROR_MARGIN = 0.106;
+constexpr double HAS_BALL_DISTANCE_ERROR_MARGIN = 0.1;
```
Comment (HaicoDorenbos) [inline src/stp/new_constants/ControlConstants.cpp]: Explain these variables with more documentation

File: src/stp/new_plays/AggressiveFormation.cpp [hunk 14]
```diff
@@ -25,7 +25,7 @@ AggressiveFormation::AggressiveFormation() : Play() {
         std::make_unique<role::BallAvoider>(role::BallAvoider("offender_3"))};
 }
 
-uint8_t AggressiveFormation::score(world_new::World* world) noexcept { return 2; }
+uint8_t AggressiveFormation::score(world_new::World* world) noexcept { return 100; }
```
Comment (HaicoDorenbos) [inline src/stp/new_plays/AggressiveFormation.cpp:34]: Make a constant class variable for 100 and explain the meaning of it.

File: src/stp/new_constants/ControlConstants.cpp [hunk 15]
```diff
@@ -36,11 +36,11 @@ constexpr double MAX_DRIBBLER_CMD = 31;
...
 
 // GTP Constants
-constexpr double GO_TO_POS_ERROR_MARGIN = 0.018;
-constexpr double GO_TO_POS_ANGLE_ERROR_MARGIN = 0.019;
+constexpr double GO_TO_POS_ERROR_MARGIN = 0.01;
+constexpr double GO_TO_POS_ANGLE_ERROR_MARGIN = 0.01;
```
Comment (HaicoDorenbos) [inline src/stp/new_constants/ControlConstants.cpp]: These constants should be properly documented

File: src/stp/new_plays/DefendPass.cpp [hunk 16]
```diff
@@ -35,7 +35,7 @@ DefendPass::DefendPass() : Play() {
         std::make_unique<role::Formation>(role::Formation("offender_2"))};
 }
 
-uint8_t DefendPass::score(world_new::World *world) noexcept { return 90; }
+uint8_t DefendPass::score(world_new::World *world) noexcept { return 100; }
```
Comment (HaicoDorenbos) [inline src/stp/new_plays/DefendPass.cpp:38]: Again, make a constant for this class variable.

File: src/stp/new_plays/GenericPass.cpp [hunk 17]
```diff
@@ -41,21 +41,21 @@ GenericPass::GenericPass() : Play() {
                                                                                  std::make_unique<role::Halt>(role::Halt("halt_7"))};
 }
 
-uint8_t GenericPass::score(world_new::World* world) noexcept { return 50; }
+uint8_t GenericPass::score(world_new::World* world) noexcept { return 10; }
```
Comment (HaicoDorenbos) [inline src/stp/new_plays/GenericPass.cpp]: Class constant variable...

File: src/stp/new_plays/DefendPass.cpp (hunk 16 above)
Comment (timovdk) [reply src/stp/new_plays/DefendPass.cpp:38]: No because the scoring will be fixed in  a future PR

File: src/stp/new_plays/AggressiveFormation.cpp (hunk 14 above)
Comment (timovdk) [reply src/stp/new_plays/AggressiveFormation.cpp:34]: No because the scoring will be fixed in a future PR

File: src/stp/new_plays/GenericPass.cpp (hunk 17 above)
Comment (timovdk) [reply src/stp/new_plays/GenericPass.cpp]: No because the scoring will be fixed in a future PR

File: src/stp/new_plays/GenericPass.cpp [hunk 18]
```diff
@@ -41,21 +41,21 @@ GenericPass::GenericPass() : Play() {
...
 
 void GenericPass::calculateInfoForRoles() noexcept {
     // Keeper
     stpInfos["keeper"].setPositionToShootAt(Vector2{0.0, 0.0});
     stpInfos["keeper"].setEnemyRobot(world->getWorld()->getRobotClosestToBall(world_new::them));
 
-    const Vector2 passingPosition = calculatePassLocation();
+    const Vector2 passingPosition = Vector2();///*Vector2(-field.getFieldLength()/2, -field.getFieldWidth()/2);//*/calculatePassLocation();
```
Comment (HaicoDorenbos) [inline src/stp/new_plays/GenericPass.cpp]: Do not create something that does not exist yet. Use optionals or pointers instead if you want to define a variable at a later time.

Comment (timovdk) [reply src/stp/new_plays/GenericPass.cpp]: Why would it not exist? Vector2 is default initialized to (0,0) so it is a completely valid thing to do

File: include/roboteam_ai/stp/new_constants/ControlConstants.h [hunk 19]
```diff
@@ -7,24 +7,21 @@
...
-extern const double DEFAULT_KICK_POWER;
 extern const double MAX_POWER_KICK_DISTANCE;
 extern const double MAX_POWER_CHIP_DISTANCE;
 
 // Dribbler constants
 extern const double TURN_ON_DRIBBLER_DISTANCE;
 
 // Team constants TODO: Maybe this should be in a different constants file
-inline constexpr size_t MAX_ROBOT_COUNT = 11;
+inline constexpr int MAX_ROBOT_COUNT = 11;
```
Comment (Yuhanun) [inline include/roboteam_ai/stp/new_constants/ControlConstants.h]: uint8_t

File: include/roboteam_ai/stp/new_roles/BallReflector.h [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+
+namespace rtt::ai::stp::role {
+
+class BallReflector : public Role {
+public:
+    /**
+     * Ctor that sets the name of the role and creates a statemachine of tactics
+     * @param name name of the role
+     */
+ BallReflector(std::string name);
```
Comment (Yuhanun) [inline include/roboteam_ai/stp/new_roles/BallReflector.h:18]:
```suggestion
 BallReflector(std::string_view name);
```

File: include/roboteam_ai/world_new/World.hpp [hunk 21]
```diff
@@ -79,7 +79,7 @@ class World {
      *
      * Undefined behavior may occur if feedback is used after being passed to this function
      */
-    void updateFeedback(uint8_t robotId, proto::RobotFeedback &feedback);
+    void updateFeedback(std::unordered_map<uint8_t, proto::RobotFeedback> feedback);
```
Comment (Yuhanun) [inline include/roboteam_ai/world_new/World.hpp:82]: Why this change? Do we do batch updating now?

File: src/interface/widgets/widget.cpp [hunk 22]
```diff
@@ -23,12 +23,14 @@ namespace rtt::ai::interface {
...
-        drawFieldLines(field, painter);
+        if(field.has_value()) {
+            calculateFieldSizeFactor(field.value());
+            drawBackground(painter);
+            drawFieldHints(field.value(), painter);
+            drawFieldLines(field.value(), painter);
+        }
 
         QString s;
         s.fromStdString("We have " + std::to_string(world->getUs().size()) + " robots");
```
Comment (Yuhanun) [inline src/interface/widgets/widget.cpp]:
```suggestion
        auto s = QString::fromStdString("We have " + std::to_string(world->getUs().size()) + " robots");
```

File: src/stp/Skill.cpp [hunk 23]
```diff
@@ -84,16 +84,16 @@ void Skill::limitAngularVel() noexcept {
     // Limit the angular velocity when the robot has the ball by setting the target angle in small steps
     // TODO: Might want to limit on the robot itself
     if (robot->hasBall() && command.use_angle()) {
-        double angleRate = 0.2 * M_PI; // Angle increment per tick TODO: TUNE
+        constexpr double ANGLE_RATE = 0.1 * M_PI; // Angle increment per tick TODO: TUNE
```
Comment (Yuhanun) [inline src/stp/Skill.cpp]: constants.h?

File: src/stp/new_constants/ControlConstants.cpp [hunk 24]
```diff
@@ -2,67 +2,65 @@
...
 constexpr double FUZZY_FALSE = 0;
 constexpr double FUZZY_MARGIN = 0.1;
 constexpr double FUZZY_DEFAULT_CUTOFF = 127;
 
-// Distance constants
-const double DISTANCE_TO_ROBOT_CLOSE = ROBOT_RADIUS; // TODO: TUNE
-const double DISTANCE_TO_ROBOT_FAR = 5 * ROBOT_RADIUS; // TODO: TUNE
+/// Distance constants
+const double DISTANCE_TO_ROBOT_CLOSE = ROBOT_RADIUS;
+const double DISTANCE_TO_ROBOT_FAR = 5 * ROBOT_RADIUS;
```
Comment (Yuhanun) [inline src/stp/new_constants/ControlConstants.cpp]:
```suggestion
constexpr double DISTANCE_TO_ROBOT_CLOSE = ROBOT_RADIUS;
constexpr double DISTANCE_TO_ROBOT_FAR = 5 * ROBOT_RADIUS;
```

File: src/stp/new_constants/ControlConstants.cpp [hunk 25]
```diff
@@ -2,67 +2,65 @@
...
 constexpr double ROBOT_CLOSE_TO_POINT = 0.2;
 
-// Keeper constants
+/// Keeper constants
 const double DISTANCE_FROM_GOAL_CLOSE = 2 * control_constants::ROBOT_RADIUS;
```
Comment (Yuhanun) [inline src/stp/new_constants/ControlConstants.cpp]:
```suggestion
constexpr double DISTANCE_FROM_GOAL_CLOSE = 2 * control_constants::ROBOT_RADIUS;
```

File: src/stp/new_constants/ControlConstants.cpp [hunk 26]
```diff
@@ -2,67 +2,65 @@
...
 
-// GameState constants
+/// GameState constants
 const double AVOID_BALL_DISTANCE = 0.55;
```
Comment (Yuhanun) [inline src/stp/new_constants/ControlConstants.cpp]:
```suggestion
constexpr double AVOID_BALL_DISTANCE = 0.55;
```

File: src/stp/new_plays/Attack.cpp [hunk 27]
```diff
@@ -38,7 +38,7 @@ Attack::Attack() : Play() {
                                                                                  std::make_unique<role::Defender>(role::Defender("defender_3"))};
 }
 
-uint8_t Attack::score(world_new::World *world) noexcept { return 110; }
+uint8_t Attack::score(world_new::World *world) noexcept { return 80; }
```
Comment (Yuhanun) [inline src/stp/new_plays/Attack.cpp:41]: Why are these constants?... shouldn't it weigh in wehther we have ball or not etc?...

File: src/stp/new_plays/AttackingPass.cpp [hunk 28]
```diff
@@ -189,13 +189,15 @@ bool AttackingPass::isValidPlayToKeep(world_new::World *world) noexcept {
 }
 
 bool AttackingPass::passFinished() noexcept {
+    //TODO: fix this condition
     if(stpInfos["receiver"].getRobot() && stpInfos["receiver"].getRobot()->get()->getDistanceToBall() < 0.5) {
         return true;
     }
     return false;
```
Comment (Yuhanun) [inline src/stp/new_plays/AttackingPass.cpp]:
```suggestion
    return stpInfos["receiver"].getRobot() && stpInfos["receiver"].getRobot()->get()->getDistanceToBall() < 0.5;
```

File: src/stp/new_tactics/KickAtPos.cpp [hunk 29]
```diff
@@ -4,9 +4,9 @@
 
 #include "stp/new_tactics/KickAtPos.h"
 
+#include <include/roboteam_ai/utilities/Constants.h>
```
Comment (Yuhanun) [inline src/stp/new_tactics/KickAtPos.cpp]:
```suggestion
#include <utilities/Constants.h>
```

File: src/stp/new_tactics/KickAtPos.cpp [hunk 30]
```diff
@@ -54,9 +50,9 @@ double KickAtPos::determineKickForce(double distance, KickChipType desiredBallSp
     // TODO: TUNE these factors need tuning
     // Increase these factors to decrease kick velocity
     // Decrease these factors to increase kick velocity
-    const double TARGET_FACTOR{1.4};
+    const double TARGET_FACTOR{2.65};
```
Comment (Yuhanun) [inline src/stp/new_tactics/KickAtPos.cpp]:
```suggestion
    constexpr double TARGET_FACTOR{2.65};
```

File: src/stp/new_tactics/KickAtPos.cpp [hunk 31]
```diff
@@ -54,9 +50,9 @@ double KickAtPos::determineKickForce(double distance, KickChipType desiredBallSp
...
     const double GRSIM_TARGET_FACTOR{1.65};
-    const double PASS_FACTOR{1.2};
+    const double PASS_FACTOR{1.45};
```
Comment (Yuhanun) [inline src/stp/new_tactics/KickAtPos.cpp]:
```suggestion
    constexpr double PASS_FACTOR{1.45};
```

File: src/stp/new_tactics/KickAtPos.cpp [hunk 32]
```diff
@@ -54,9 +50,9 @@ double KickAtPos::determineKickForce(double distance, KickChipType desiredBallSp
     // TODO: TUNE these factors need tuning
     // Increase these factors to decrease kick velocity
     // Decrease these factors to increase kick velocity
-    const double TARGET_FACTOR{1.4};
+    const double TARGET_FACTOR{2.65};
     const double GRSIM_TARGET_FACTOR{1.65};
```
Comment (Yuhanun) [inline src/stp/new_tactics/KickAtPos.cpp]:
```suggestion
    constexpr double GRSIM_TARGET_FACTOR{1.65};
```

File: src/stp/new_tactics/KickAtPos.cpp [hunk 33]
```diff
@@ -54,9 +50,9 @@ double KickAtPos::determineKickForce(double distance, KickChipType desiredBallSp
...
-    const double PASS_FACTOR{1.2};
+    const double PASS_FACTOR{1.45};
     const double GRSIM_PASS_FACTOR{1.45};
```
Comment (Yuhanun) [inline src/stp/new_tactics/KickAtPos.cpp]:
```suggestion
    constexpr double GRSIM_PASS_FACTOR{1.45};
```

File: src/stp/new_tactics/Receive.cpp [hunk 34]
```diff
@@ -33,7 +33,7 @@ std::optional<StpInfo> Receive::calculateInfoForSkill(StpInfo const &info) noexc
     skillStpInfo.setAngle(calculateAngle(info.getRobot().value(), info.getBall().value()));
 
     // If ball is close to robot, turn on dribbler
-    skillStpInfo.setDribblerSpeed(determineDribblerSpeed(info.getRobot().value()));
+    if(skillStpInfo.getRobot()->get()->getDistanceToBall() <= control_constants::TURN_ON_DRIBBLER_DISTANCE*5) skillStpInfo.setDribblerSpeed(100);
```
Comment (Yuhanun) [inline src/stp/new_tactics/Receive.cpp]:
```suggestion
    if(skillStpInfo.getRobot()->get()->getDistanceToBall() <= control_constants::TURN_ON_DRIBBLER_DISTANCE*5) {
        skillStpInfo.setDribblerSpeed(100);
    }
```

File: src/stp/new_tactics/Receive.cpp [hunk 35]
```diff
@@ -54,7 +54,14 @@ bool Receive::isEndTactic() noexcept {
     return true;
 }
 
-double Receive::calculateAngle(const world_new::view::RobotView &robot, const world_new::view::BallView &ball) { return (ball->getPos() - robot->getPos()).angle(); }
+double Receive::calculateAngle(const world_new::view::RobotView &robot, const world_new::view::BallView &ball) {
+    if(robot->getDistanceToBall() <= 1.0) {
+    return robot->getAngle();
```
Comment (Yuhanun) [inline src/stp/new_tactics/Receive.cpp]:
```suggestion
        return robot->getAngle();
```

File: src/stp/new_tactics/Receive.cpp [hunk 36]
```diff
@@ -54,7 +54,14 @@ bool Receive::isEndTactic() noexcept {
...
+    }
+    else{
```
Comment (Yuhanun) [inline src/stp/new_tactics/Receive.cpp]:
```suggestion
    } else {
```

File: src/utilities/IOManager.cpp [hunk 37]
```diff
@@ -91,6 +92,13 @@ const proto::SSL_Referee &IOManager::getRefereeData() {
     return *msg;
 }
 
+const std::unordered_map<uint8_t, proto::RobotFeedback> &IOManager::getFeedbackDataMap() {
+    std::lock_guard<std::mutex> lock(robotFeedbackMutex);
+    auto fbm = new std::unordered_map<uint8_t, proto::RobotFeedback>;
```
Comment (Yuhanun) [inline src/utilities/IOManager.cpp]: This is a memory leak

Comment (Yuhanun) [reply src/utilities/IOManager.cpp]: I think you want to return the internal unordered map instead of this.

File: src/utilities/IOManager.cpp [hunk 38]
```diff
@@ -91,6 +92,13 @@ const proto::SSL_Referee &IOManager::getRefereeData() {
     return *msg;
 }
 
+const std::unordered_map<uint8_t, proto::RobotFeedback> &IOManager::getFeedbackDataMap() {
```
Comment (Yuhanun) [inline src/utilities/IOManager.cpp]: This whole function is a hack, either return a copy of the internal world, aka remove const&, or return a const& to the internal map, dont heap allocate and then return a reference to the heap allocated object

**Review by Yuhanun — CHANGES_REQUESTED** [review]
Memory leak _has_ to be fixed.

File: include/roboteam_ai/world_new/World.hpp (hunk 21 above)
Comment (timovdk) [reply include/roboteam_ai/world_new/World.hpp:82]: At the time this was the quickest fix, but we can change to id-based updating if you like. That would mean that there will be a loop in application manager that will update 11 robots on id, aka looking up the mapped value in IOManager and using that as input for the map that is in World

File: src/stp/new_plays/Attack.cpp (hunk 27 above)
Comment (timovdk) [reply src/stp/new_plays/Attack.cpp:41]: The scoring is TBD

File: test/World_newTests/WhichRobotHasBallTest.cpp [hunk 39]
```diff
@@ -41,40 +41,27 @@ TEST(World_newTest, WhichRobotHasBallTest) {
...
-    worldInstance->updateWorld(msgNoBall);
-
-    auto world = worldInstance->getWorld();
-    assert(world->getUs().size() == 2);
-    assert(world->getThem().size() == 1);
-    assert(!world->getBall().has_value());
-    std::optional<rtt::world_new::view::RobotView> robot = world->whichRobotHasBall(rtt::world_new::Team::us);
-    EXPECT_TRUE(!robot.has_value());
-
-    /** Test 2 : Two yellow robots (1 & 2), two blue robots (3 & 4), one ball **/
```
Comment (RobotJente) [inline test/World_newTests/WhichRobotHasBallTest.cpp:57]: Why was this removed?

Comment (timovdk) [reply test/World_newTests/WhichRobotHasBallTest.cpp:57]: We always have a ball now whenever we create a ball from a proto message after the fix in Ball.cpp. (the ball staying with a robot when there is no vision)

**Review by RobotJente — APPROVED** [review]
Go RoboTimo! En YohmaYordi!

File: src/interface/widgets/widget.cpp [hunk 40]
```diff
@@ -263,7 +264,7 @@ void Visualizer::drawFieldHints(const world::Field &field, QPainter &painter) {
         // draw us
         for (auto const &robot : world->getUs()) {
             std::string role{};
-            if (rolesForRobots.find(robot->getId()) != rolesForRobots.end()) {
+            if (robot && rolesForRobots.find(robot->getId()) != rolesForRobots.end()) {
```
Comment (Yuhanun) [inline src/interface/widgets/widget.cpp]: how is this condition not redundant?

File: src/stp/Tactic.cpp [hunk 41]
```diff
@@ -43,18 +41,17 @@ Status Tactic::update(StpInfo const &info) noexcept {
         return Status::Waiting;
     }
 
+    // the tactic will not be reset if it's the first skill
+    if ((skills.current_num() != 0 && shouldTacticReset(skill_info.value()))) {
```
Comment (Yuhanun) [inline src/stp/Tactic.cpp]:
```suggestion
    if (skills.current_num() != 0 && shouldTacticReset(skill_info.value())) {
```

**Review by Yuhanun — APPROVED** [review]
lgtm

File: src/stp/new_constants/ControlConstants.cpp (hunk 15 above)
Comment (Yuhanun) [reply src/stp/new_constants/ControlConstants.cpp]: I mean they're pretty clear but documentation on these would be nice.

File: src/interface/widgets/widget.cpp (hunk 40 above)
Comment (timovdk) [inline src/interface/widgets/widget.cpp]:
```suggestion
            if (rolesForRobots.find(robot->getId()) != rolesForRobots.end()) {
```
